### PR TITLE
Tip 3 draft

### PR DESCRIPTION
**Pull request by fmaguire**
**Did you add yourself as a [contributor](https://github.com/Benjamin-Lee/deep-rules/blob/master/contributors.md) if this is your first contribution?**

- [x] Yes, I added myself or am already a contributor

**Any more details?**

Draft of tip 3, with points from #42 #82 #44 #52 #63 

Could be expanded to cover more of #15 

Wasn't sure how much of #22 to add without being redundant to Tip 4.

### CODE REVIEW

File: content/05.dl-complexities.md [hunk 1]
```diff
@@ -1 +1,25 @@
 ## Tip 3: Understand the complexities of training deep neural networks {#complexities}
+
+Correctly training deep neural networks is a non-trivial process.
+There are many different options and potential pitfalls at every stage.
+To get good results you have to expect to train many networks with
+different tweaks and modifications.
+Despite improved framework ease-of-use and on-demand cloud computing resources
+this means DL can be very demanding, requiring significant
+infrastructure and patitence to complete.
```
Comment (rasbt) [inline content/05.dl-complexities.md]: "patitence" => "patience"

File: content/05.dl-complexities.md [hunk 2]
```diff
@@ -1 +1,25 @@
...
+The experimentation inherent to DL is often noisy (requiring repetition) and
+represent a significant organisational challenge.
```
Comment (rasbt) [inline content/05.dl-complexities.md]: "organisational" => "organizational"

(sorry, don't want to be nit-picky, but we used AE elsewhere)

File: content/05.dl-complexities.md [hunk 3]
```diff
@@ -1 +1,25 @@
...
+All code, parameters, and results must be carefully corralled using good
+coding practices (e.g. version control, continuous integration etc.) in order to
```
Comment (rasbt) [inline content/05.dl-complexities.md]: e.g. and etc. shouldn't be used in the same statement. I suggest

- (e.g., version control and continuous integration)

or 

- (for example, version control and continuous integration)

File: content/05.dl-complexities.md [hunk 4]
```diff
@@ -1 +1,25 @@
...
+different tweaks and modifications.
+Despite improved framework ease-of-use and on-demand cloud computing resources
+this means DL can be very demanding, requiring significant
+infrastructure and patitence to complete.
+The experimentation inherent to DL is often noisy (requiring repetition) and
+represent a significant organisational challenge.
+All code, parameters, and results must be carefully corralled using good
+coding practices (e.g. version control, continuous integration etc.) in order to 
+be effective and interpretable.
+This organisation is also key to being able to efficiently share your work
```
Comment (rasbt) [inline content/05.dl-complexities.md]: "organisation" => "organization"

File: content/05.dl-complexities.md (hunk 1 above)
Comment (fmaguire) [reply content/05.dl-complexities.md]: Fixed

File: content/05.dl-complexities.md (hunk 2 above)
Comment (fmaguire) [reply content/05.dl-complexities.md]: Not at all, totally automatic Britishism on my behalf.

File: content/05.dl-complexities.md [hunk 5]
```diff
@@ -1 +1,25 @@
...
+and to update your model as new data becomes available. 
+
+
+Similar to [Tip 4](#baselines), try to start with a relatively smaller network
```
Comment (rasbt) [inline content/05.dl-complexities.md]: Not sure if we were 100% consistent about that, but I think we referred to the reader in 3rd person before ("the researcher", "the experimenter," etc.)

File: content/05.dl-complexities.md (hunk 3 above)
Comment (fmaguire) [reply content/05.dl-complexities.md]: Fixed.

File: content/05.dl-complexities.md (hunk 5 above)
Comment (fmaguire) [reply content/05.dl-complexities.md]: Might be easiest to fix that flow/match tone when we have a draft for all sections?

File: content/05.dl-complexities.md [hunk 6]
```diff
@@ -1 +1,25 @@
...
+and increase the size and complexity as needed to prevent wasting time and 
+resources. 
+Beware of the seemingly trivial choices that are being made implicitly by 
+default settings in your framework of choice e.g. choice of optimization algorithm
+(adaptive methods are often faster but can lead to worse results [@arxiv:1705.08292]).
```
Comment (rasbt) [inline content/05.dl-complexities.md]: > "adaptive methods are often faster but can lead to worse results"

I'd suggest

"adaptive methods often lead to faster convergence during training but may lead to worse generalization performance on independent datasets" 

@arxiv:1705.08292: This was a 2017 NIPS paper, maybe it would be worth citing the peer-reviewed version:

Wilson, A. C., Roelofs, R., Stern, M., Srebro, N., & Recht, B. (2017). The marginal value of adaptive gradient methods in machine learning. In Advances in Neural Information Processing Systems (pp. 4148-4158).

A related one is https://arxiv.org/abs/1712.07628 by Keskar & Socher, but not sure if this has been published somewhere

File: content/05.dl-complexities.md (hunk 5 above)
Comment (rasbt) [reply content/05.dl-complexities.md]: I agree. Right now, these things are a bit difficult to address without having the context of all chapters!

File: content/05.dl-complexities.md (hunk 6 above)
Comment (fmaguire) [reply content/05.dl-complexities.md]: Think I managed to add the json format properly... guess I'll find out in a second once travis finishes...

Comment (fmaguire) [reply content/05.dl-complexities.md]: Updated to Wilson citation from the NeurIPS proceedings

File: content/05.dl-complexities.md (hunk 5 above)
Comment (rasbt) [reply content/05.dl-complexities.md]: Sounds good. Like you suggested before, it may be better to wait until we have all the tips outlined in a first draft before we smooth it out to make it consistent :)

File: content/05.dl-complexities.md (hunk 6 above)
Comment (agitter) [reply content/05.dl-complexities.md]: For NeurIPS proceedings, URL citations are an option so that you don't have to manually provide the JSON: `url:https://papers.nips.cc/paper/7003-the-marginal-value-of-adaptive-gradient-methods-in-machine-learning`

It doesn't work well with every conference proceedings site, but this paper looks accurate.

File: content/05.dl-complexities.md [hunk 7]
```diff
@@ -1 +1,25 @@
 ## Tip 3: Understand the complexities of training deep neural networks {#complexities}
+
+Correctly training deep neural networks is a non-trivial process.
+There are many different options and potential pitfalls at every stage.
+To get good results you have to expect to train many networks with
```
Comment (agitter) [inline content/05.dl-complexities.md]: Can you please put a single sentence on each line instead of using hard line wrapping?  That makes it easier to track and comment on each sentence.

Comment (fmaguire) [reply content/05.dl-complexities.md]: Sure, that makes sense. Done.

File: content/manual-references.json [hunk 8]
```diff
@@ -18,5 +18,74 @@
       "family": "Himmelstein"
     }
   ]
-}
-]
\ No newline at end of file
+},
+{
```
Comment (agitter) [inline content/manual-references.json]: If the URL citation is working now, you can please revert this file to remove the manual reference.  Thanks.

Comment (fmaguire) [reply content/manual-references.json]: Sure thing, done now.

File: content/05.dl-complexities.md [hunk 9]
```diff
@@ -1 +1,14 @@
 ## Tip 3: Understand the complexities of training deep neural networks {#complexities}
+
+Correctly training deep neural networks is a non-trivial process.
+There are many different options and potential pitfalls at every stage.
+To get good results you have to expect to train many networks with different tweaks and modifications.
+Despite improved framework ease-of-use and on-demand cloud computing resources this means DL can be very demanding, requiring significant infrastructure and patience to complete.
+The experimentation inherent to DL is often noisy (requiring repetition) and represents a significant organizational challenge.
+All code, parameters, and results must be carefully corralled using good coding practices (for example, version control, continuous integration etc.) in order to be effective and interpretable.
```
Comment (agitter) [inline content/05.dl-complexities.md]: I'm curious about examples of continuous integration with DL.  Is there anything we could refer to?  Given that model training and performance can depend on things like the hardware and versions of required packages, I am interested in best practices here and think our readers would be as well.

File: content/05.dl-complexities.md [hunk 10]
```diff
@@ -1 +1,14 @@
...
+This organization is also key to being able to efficiently share your work and to update your model as new data becomes available. 
+
```
Comment (agitter) [inline content/05.dl-complexities.md]: Extra linebreak here

File: content/05.dl-complexities.md (hunk 9 above)
Comment (rasbt) [reply content/05.dl-complexities.md]: We definitely to mention non-deterministic algorithm choices in CUDA & cuDNN. I.e., by default in e.g., TensorFlow and PyTorch, running code on the GPU (as opposed to CPU) is non-deterministic usually (esp. convolutions). This is because cuDNN has many different approximations of convolutions implemented (i.e., all the winnograd stuff and derivates thereof approximating convolutions via fast fourier transforms). So, by default, if you run GPU code, the "best" algorithm is chosen at runtime and may vary from run to run (and can also change during runtime I think). Hence, it's best practice for reproducibility to set `torch.backends.cudnn.deterministic = True` in PyTorch, for example (forgot how to do that in TensorFlow). In practice, it's also often faster as the "guessing" by cuDNN is not that good (heard that from someone at NVIDIA as well when I reported it). 

Note that setting random seeds will only affect things like deterministic shuffling and weight initialization, the algorithm determinism has to be set on top of that. I think this is important to mention. We could reference the cuDNN technical manuals for this: https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#reproducibility

File: content/05.dl-complexities.md [hunk 11]
```diff
@@ -1 +1,14 @@
 ## Tip 3: Understand the complexities of training deep neural networks {#complexities}
+
+Correctly training deep neural networks is a non-trivial process.
+There are many different options and potential pitfalls at every stage.
+To get good results you have to expect to train many networks with different tweaks and modifications.
```
Comment (rasbt) [inline content/05.dl-complexities.md]: > "with different tweaks and modifications." 

 "tweaks and modifications" sounds good, but we should also say "settings" because changing things like the learning rate is not necessarily a tweak or modification (I think of tweaks and modifications more of larger changes, like adding dropout, batchnorm, varying the number of layers, and so forth)

File: content/05.dl-complexities.md [hunk 12]
```diff
@@ -1 +1,14 @@
...
+Despite improved framework ease-of-use and on-demand cloud computing resources this means DL can be very demanding, requiring significant infrastructure and patience to complete.
```
Comment (rasbt) [inline content/05.dl-complexities.md]: I think we need to add a comparison here, because that's of course very relative. Just attended a seminar last week where someone ran simulations on 4096 CPUs that took months to complete ... the throughput was 1 petaflop, which could have been easily achieved via a handful of GPUs -- what I am trying to say is that the infrastructure is relatively demanding compared to maybe traditional machine learning on small datasets. Also, traditional MCMC or all the molecular dynamics research is much more expensive than DL (e.g., I would say DL is currently probably the cheapest method for protein structure refinement compared to traditional methods :P)

**Comment by rasbt** [comment]
> @rasbt did you have any other feedback?

I just added some additional comments we may want to address, i.e., 

- including sth like "settings" when talking about "requiring modifications"
- adding reference for comparison when saying that DL is demanding
- the reproducibility aspect with GPU computing

we could add this later, but since everything could be addressed on 0.5 -2 sentences, I think it's organizationally easier to do that before we merge (if you agree with these points, of course). Otherwise and overall, it looks pretty good to me :)

**Comment by agitter** [comment]
Yes, I agree with your points above.  Because they are fairly specific suggestions, I also think we can make these changes before merging.

File: content/05.dl-complexities.md (hunk 12 above)
Comment (fmaguire) [reply content/05.dl-complexities.md]: I couldn't really find a good general comparison but I did find a paper that was looking at scaling and computational performance.  We don't need to say necessarily it is WORSE than other methods to have a strong case that it can still require significant infrastructure etc.

File: content/05.dl-complexities.md (hunk 9 above)
Comment (fmaguire) [reply content/05.dl-complexities.md]: Couldn't really find a good CI example specifically for DL usage instead of writing new modules/frameworks.   

Good call on the non-deterministic/deterministic backend algorithms part, I'd forgotten totally about that.
